File: js/cli/src/ui.ts
import chalk from 'chalk';
import ora, { Ora } from 'ora';

// Human-friendly output (with colors and symbols)
export function success(message: string): void {
  console.log(chalk.green('✓'), message);
}

export function error(message: string): void {
  console.log(chalk.red('✗'), message);
}

export function info(message: string): void {
  console.log(chalk.blue('ℹ'), message);
}

export function warn(message: string): void {
  console.log(chalk.yellow('⚠'), message);
}

export function spinner(text: string): Ora {
  return ora(text).start();
}

export function printHeader(title: string): void {
  console.log();
  console.log(chalk.bold.cyan(title));
  console.log(chalk.gray('─'.repeat(title.length)));
  console.log();
}

export function printKeyValue(key: string, value: string): void {
  console.log(chalk.gray(`${key}:`), chalk.white(value));
}

export function printSection(title: string): void {
  console.log();
  console.log(chalk.bold(title));
}

// LLM-friendly output (structured, compact, no colors)
export function llmSection(title: string): void {
  try {
    console.log();
    console.log(title.toUpperCase());
    console.log();
  } catch (err: any) {
    if (err.code === 'EPIPE') {
      process.exit(0);
    }
    throw err;
  }
}

export function llmKeyValue(key: string, value: string): void {
  try {
    console.log(`${key}: ${value}`);
  } catch (err: any) {
    if (err.code === 'EPIPE') {
      process.exit(0);
    }
    throw err;
  }
}

export function llmTable(headers: string[], rows: string[][]): void {
  if (rows.length === 0) return;

  try {
    const colWidths = headers.map((h, i) => {
      const maxDataWidth = Math.max(...rows.map(r => (r[i] || '').length));
      return Math.max(h.length, maxDataWidth);
    });

    const separator = '|' + colWidths.map(w => '-'.repeat(w + 2)).join('|') + '|';
    const headerRow = '|' + headers.map((h, i) => ` ${h.padEnd(colWidths[i])} `).join('|') + '|';

    console.log(headerRow);
    console.log(separator);

    for (const row of rows) {
      const rowStr = '|' + row.map((cell, i) => ` ${(cell || '').padEnd(colWidths[i])} `).join('|') + '|';
      console.log(rowStr);
    }
  } catch (err: any) {
    if (err.code === 'EPIPE') {
      process.exit(0);
    }
    throw err;
  }
}

export function llmList(items: Array<{ key: string; value: string; indent?: number }>): void {
  for (const item of items) {
    const indent = '  '.repeat(item.indent || 0);
    console.log(`${indent}${item.key}: ${item.value}`);
  }
}

File: rust/cli-rs/src/ui.rs
use colored::Colorize;
use std::io::Write;

/// Human-friendly output with colors and symbols
#[allow(dead_code)]
pub struct HumanFormatter;

#[allow(dead_code)]
impl HumanFormatter {
    pub fn success(message: &str) {
        let _ = writeln!(std::io::stdout(), "{} {}", "✓".green(), message);
    }

    pub fn error(message: &str) {
        let _ = writeln!(std::io::stderr(), "{} {}", "✗".red(), message);
    }

    pub fn info(message: &str) {
        let _ = writeln!(std::io::stdout(), "{} {}", "ℹ".blue(), message);
    }

    pub fn section(title: &str) {
        let _ = writeln!(std::io::stdout());
        let _ = writeln!(std::io::stdout(), "{}", title.bold());
    }

    pub fn key_value(key: &str, value: &str) {
        let _ = writeln!(std::io::stdout(), "  {}: {}", key.dimmed(), value);
    }
}

/// LLM-friendly output (structured, compact, no colors)
pub struct LlmFormatter;

impl LlmFormatter {
    pub fn section(title: &str) {
        // Handle EPIPE gracefully by ignoring broken pipe errors
        if let Err(e) = std::io::Write::write_all(
            &mut std::io::stdout(),
            format!("\n{}\n\n", title.to_uppercase()).as_bytes(),
        ) {
            if e.kind() == std::io::ErrorKind::BrokenPipe {
                std::process::exit(0);
            }
        }
    }

    pub fn key_value(key: &str, value: &str) {
        // Handle EPIPE gracefully by ignoring broken pipe errors
        if let Err(e) = std::io::Write::write_all(
            &mut std::io::stdout(),
            format!("{}: {}\n", key, value).as_bytes(),
        ) {
            if e.kind() == std::io::ErrorKind::BrokenPipe {
                std::process::exit(0);
            }
        }
    }

    pub fn table(headers: &[&str], rows: &[Vec<String>]) {
        if rows.is_empty() {
            return;
        }

        // Calculate column widths
        let col_widths: Vec<usize> = headers
            .iter()
            .enumerate()
            .map(|(i, h)| {
                let max_data_width = rows
                    .iter()
                    .map(|r| r.get(i).map(|s| s.len()).unwrap_or(0))
                    .max()
                    .unwrap_or(0);
                h.len().max(max_data_width)
            })
            .collect();

        // Print header
        let _ = write!(std::io::stdout(), "|");
        for (i, header) in headers.iter().enumerate() {
            let _ = write!(
                std::io::stdout(),
                " {:<width$} |",
                header,
                width = col_widths[i]
            );
        }
        let _ = writeln!(std::io::stdout());

        // Print separator
        let _ = write!(std::io::stdout(), "|");
        for width in &col_widths {
            let _ = write!(std::io::stdout(), "{}", "-".repeat(width + 2));
            let _ = write!(std::io::stdout(), "|");
        }
        let _ = writeln!(std::io::stdout());

        // Print rows
        for row in rows {
            let _ = write!(std::io::stdout(), "|");
            for (i, cell) in row.iter().enumerate() {
                let _ = write!(
                    std::io::stdout(),
                    " {:<width$} |",
                    cell,
                    width = col_widths[i]
                );
            }
            let _ = writeln!(std::io::stdout());
        }
    }
}
